### PR TITLE
Add Pytorch support for Cum Op

### PR DESCRIPTION
**Pull request by HarshvirSandhu**
## Description
This PR can be used as an example for implementing `Op`s in PyTorch

## Related Issue

- [x] Related to #821 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

cc @ricardoV94

### CODE REVIEW

File: pytensor/link/pytorch/dispatch/extra_ops.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+from pytensor.link.pytorch.dispatch.basic import pytorch_funcify
+from pytensor.tensor.extra_ops import CumOp
+
+
+@pytorch_funcify.register(CumOp)
+def pytorch_funcify_Cumop(op, **kwargs):
+    dim = op.axis
+    mode = op.mode
+
+    def cumop(x, dim=dim, mode=mode):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/extra_ops.py:12]: This is not needed, the returned functions are never called by the user
```suggestion
    def cumop(x):
```

File: tests/link/pytorch/test_extra_ops.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+
+def test_pytorch_CumOp():
+    """Test PyTorch conversion of the `CumOp` `Op`."""
+
+    # Create a symbolic input for the first input of `CumOp`
+    a = pt.matrix("a")
+
+    # Create test value tag for a
+    a.tag.test_value = np.arange(9, dtype=config.floatX).reshape((3, 3))
```
Comment (ricardoV94) [inline tests/link/pytorch/test_extra_ops.py:17]: No need for test values and tags. We're planning to deprecate that functionality as well

File: tests/link/pytorch/test_extra_ops.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    # Create a PyTensor `FunctionGraph`
+    fgraph = FunctionGraph([a], [out])
+
+    # Pass the graph and inputs to the testing function
+    compare_pytorch_and_py(fgraph, [get_test_value(i) for i in fgraph.inputs])
+
+    # For the second mode of CumOp
+    out = pt.cumprod(a, axis=1)
+    fgraph = FunctionGraph([a], [out])
+    compare_pytorch_and_py(fgraph, [get_test_value(i) for i in fgraph.inputs])
```
Comment (ricardoV94) [inline tests/link/pytorch/test_extra_ops.py:31]: Here just pass the test values (instead of adding them as tags and then retrieving them)

File: tests/link/pytorch/test_extra_ops.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    """Test PyTorch conversion of the `CumOp` `Op`."""
+
+    # Create a symbolic input for the first input of `CumOp`
+    a = pt.matrix("a")
+
+    # Create test value tag for a
+    a.tag.test_value = np.arange(9, dtype=config.floatX).reshape((3, 3))
+
+    # Create the output variable
+    out = pt.cumsum(a, axis=0)
```
Comment (ricardoV94) [inline tests/link/pytorch/test_extra_ops.py:20]: Test `axis=None` and axis=tuple(...)` if supported by the original Op. If tuple is allowed make sure you have more dimensions (3,) and only test 2 of them. The axis can be parametrized (prod and add as well) instead of adding more conditions inside the test

**Comment by ricardoV94** [comment]
Can you extend the example in the documentation page on implementing custom JAX/NUMBA Ops to mention PyTorch and include this example as well? 

Perhaps you can use some fancy tab to select among the different modes. Is that supported @OriolAbril ?

**Comment by OriolAbril** [comment]
Not here as of now, you'd have to add an extra extension for tabs. If you'll only want tabs, then it is probably best to use https://sphinx-tabs.readthedocs.io/en/latest/, if using things like grids, dropdowns, icons... somewhere else in addition to tabs here seems a future possibility then https://sphinx-design.readthedocs.io/en/sbt-theme/ is probably best. Both should only require being added as dependencies to the doc env and adding them to the `extensions` varialbe in `conf.py`, no further configuration

**Comment by ricardoV94** [comment]
Thanks @OriolAbril either of those seems perfect. Any preference?

File: tests/link/pytorch/test_extra_ops.py (hunk 4 above)
Comment (HarshvirSandhu) [reply tests/link/pytorch/test_extra_ops.py:20]: Tried this on the original Op.  `axis=tuple(...)` does not work and gives a `TypeError`
`axis=None` gives the output as a 1-D array

Comment (ricardoV94) [reply tests/link/pytorch/test_extra_ops.py]: The Op `__init__` doesn't seem to check explicitly for axes but it does assume it is either None or an int. Can we add a check and raise an explicit ValueError if it's not either?

File: pytensor/link/pytorch/dispatch/extra_ops.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+from pytensor.link.pytorch.dispatch.basic import pytorch_funcify
+from pytensor.tensor.extra_ops import CumOp
+
+
+@pytorch_funcify.register(CumOp)
+def pytorch_funcify_Cumop(op, **kwargs):
+    dim = op.axis
+    mode = op.mode
+
+    def cumop(x, dim=dim):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/extra_ops.py:12]: Looks good, just no need for any kwargs. The function will only ever receive the node inputs

```suggestion
    def cumop(x):
```

**Comment by OriolAbril** [comment]
> Thanks @OriolAbril either of those seems perfect. Any preference?

I use sphinx-design more because I use its other features

File: tests/link/pytorch/test_extra_ops.py (hunk 4 above)
Comment (HarshvirSandhu) [reply tests/link/pytorch/test_extra_ops.py]: Checked again, there is no error if we use `axis=(0)`, pytorch also returns the same output.
The error only comes when there are more than 1 elements in the tuple (Even `np.cumsum` gives `TypeError` in this case).

We could try adding a check and raise, but would that be needed in other Op implementations?
Since this would be used as an example, it might be complicated if a check and raise is not needed for other implementations.

Comment (OriolAbril) [reply tests/link/pytorch/test_extra_ops.py]: `(0)` is `0`, not a tuple with a 0 inside it, it would have to be `(0,)` to be a tuple with a single element inside. Does it work with `(0,)`?

Comment (HarshvirSandhu) [reply tests/link/pytorch/test_extra_ops.py]: No, it gives a TypeError

Comment (ricardoV94) [reply tests/link/pytorch/test_extra_ops.py]: Which is fine but probably gives a typeerror in an obscure place. We should raise already in the init method of the Op to save people time

File: pytensor/tensor/extra_ops.py [hunk 6]
```diff
@@ -283,8 +283,11 @@ class CumOp(COp):
     def __init__(self, axis: int | None = None, mode="add"):
         if mode not in ("add", "mul"):
             raise ValueError(f'{type(self).__name__}: Unknown mode "{mode}"')
-        self.axis = axis
-        self.mode = mode
+        if isinstance(axis, int) or axis is None:
```
Comment (ricardoV94) [inline pytensor/tensor/extra_ops.py:286]: Nitpick, it's more common to just check and raise than indenting the "correct code"

```suggestion
        if not isinstance(axis, int) or axis is None:
            ### raise error
        # usual code
```

File: doc/extending/creating_a_numba_jax_op.rst [hunk 7]
```diff
@@ -89,42 +88,76 @@ as :class:`CumsumOp`\ :class:`Op`. The difference lies in that the `mode` attrib
...
-.. code:: python
 
-   def ifelse(cond, *args, n_outs=n_outs):
-       res = jax.lax.cond(
-           cond, lambda _: args[:n_outs], lambda _: args[n_outs:], operand=None
-       )
-       return res if n_outs > 1 else res[0]
+.. tab-set::
+
+        .. tab-item:: JAX/Numba
```
Comment (ricardoV94) [inline doc/extending/creating_a_numba_jax_op.rst:119]: This is not correct for Numba, can you leave it as a separate tab with [in progress] text (and open an issue) or check the source code of the Numba implementation if you want to do it correctly?

This probably applies to all the tabbed sections, no reason to combine jax and numba, and the pre-existing snippets were JAX specific

Comment (HarshvirSandhu) [reply doc/extending/creating_a_numba_jax_op.rst:119]: I was adding a separate tab for numba and found this comment. Is there anything that should be changed in `numba_funcify_DimShuffle`?
https://github.com/pymc-devs/pytensor/blob/71592152992675274ac5d89d17a305a3a77764f9/pytensor/link/numba/dispatch/elemwise.py#L688-L690

Comment (ricardoV94) [reply doc/extending/creating_a_numba_jax_op.rst:119]: Not in the context of this PR, but we should open an issue to check if that's still a problem in the newer versions of numba. Could you do that?

**Review by ricardoV94 — APPROVED** [review]

**Review by OriolAbril — COMMENTED** [review]
all tabs look rendered correctly, only left a comment so cross references to other libraries actually work

**Review by ricardoV94 — APPROVED** [review]
